Add Prettier, Lint Staged to help maintain repo consistency

damassi opened the pull request:
Closes https://github.com/DefinitelyTyped/DefinitelyTyped/pull/34894
Addresses (part of) https://github.com/DefinitelyTyped/DefinitelyTyped/issues/26847

In an effort to encourage code-consistency across all different type packages, this PR adds [Prettier](https://prettier.io) to the repo, which on pre-commit will run against changed files, via [lint-staged](https://github.com/okonet/lint-staged). On `pre-push`, it will run `yarn test`, which will ensure that errors are caught locally before they hit CI, which decreases a lot of aggregate dev time. 

PR #34894, which was approved, seems to have been abandoned. This one addresses the feedback there as well as introduces a git-hook en-route to some of the issues raised in #26847.

Additionally, runs Prettier over a few packages.

alloy commented:
I’m personally not a fan of single-quotes, but realise it’s a common preference. I’m _only_ saying this publicly, so others can realise that what matters more to maintainers–like me–is consistency and less faffing over syntax.

Big 👍 on this.

sandersn commented:
@damassi what errors did you see with `dtslint`? `npm test` uses a runner script from `types-publisher` to run it on all changed packages, so that is actually the thing that should run before commit.

damassi commented on types/react-native/index.d.ts:
```diff
@@ -3992,7 +3992,7 @@ export class Image extends ImageBase {
     static getSize(uri: string, success: (width: number, height: number) => void, failure: (error: any) => void): any;
     static prefetch(url: string): any;
     static abortPrefetch?(requestId: number): void;
-    static queryCache?(urls: string[]): Promise<Map<string, "memory" | "disk">>;
```
Oh, I need to rebase against master; this was a result of prettier being ran over react-native as an example, but then a bunch of changes took place in master, leading me to back out applying prettier here. Will update.

damassi commented:
@sandersn - I'll give dtslint another try.

damassi commented on package.json:
```diff
@@ -23,7 +23,21 @@
     },
-    "dependencies": {}
+    "dependencies": {},
+    "husky": {
```
It's quite easy to setup in a new project:
```
git init
yarn init
yarn add -D prettier
npx mrm lint-staged
```
Then make some changes and commit.

damassi replied:
Regarding Prettier, it's fairly standard now in new codebases. If you'd like to give that a try and see how it works as you write code, install the Prettier VSCode extension; it will tap into the `prettierrc` file and as you safe will apply. It's a wonderful energy saver and I encourage checking out [the website](https://prettier.io/). Otherwise, it will format on pre-commit.

damassi commented:
@sandersn - In order for `yarn test` to run a commit has to be made so that it can checked against master. I've moved the test command to a `pre-push` hook and it works great 👍  

So the flow is now: 
- On pre-commit, format
- On  pre-push, run tests

If the tests fail, no push will happen. 

One thing worth calling out is the test runner is pretty heavy on the processor. But that's a reasonable trade-off IMO compared to running validations solely in CI and introducing additional turn-over lag due to failed builds and rebased commits.

damassi commented:
Also, thoughts on the performance alert above? (Not sure now minor could impact.)

andrewbranch commented:
@damassi the warnings a false positive—formatting changes can't appreciably change the performance of the language service; there's a fair amount of natural variation in the measurements. I might bump up the threshold at which it issues a warning, or take standard deviation into account when deciding what the threshold is. At any rate, it's just supposed to serve as a sanity check, so you can safely ignore it for changes like this.

sandersn commented on package.json:
```diff
@@ -23,7 +23,22 @@
+    "husky": {
+        "hooks": {
+            "pre-commit": "lint-staged",
+            "pre-push": "echo '\n[DefinitelyTyped] Running tests...\n' && yarn test"
```
1. I don't think this is appropriate, for a couple of reasons:
  a. The tests are slow and the test infrastructure is hard to get installed correctly. I need to fix this, but it's in a bad state now.
  b. Some people are contributing minor changes that are low-priority to them. For them, it is fine to have a slow trip to CI to find out if they have errors. They'll never have more than 1 or 2 anyway.
1. Does this command work on Windows machines? I can't remember how cmd interprets quotes and `&&`. 
2. Can you use npm instead of yarn? I don't want to assume that yarn is installed.

I think we should re-evaluate this after
1. making the current infrastructure more stable.
2. creating a fast-run mode for tests to use here -- perhaps just linting each of the changed packages, ignoring dependents.

damassi replied:
Sounds good, for now I'll keep prettier in the precommit and remove prepush. We can return to this later.

damassi commented:
@sandersn 

> Instead, we should run prettier on 10-20 packages at a time. 

I think this could work if we keep the merge pipeline consistently flowing. Some packages get updated quite frequently and we'd want to avoid merge conflicts. How do you suggest we coordinate this? 

(cc @alloy, as he may have some ideas.)

alloy commented:
Wouldn’t it be easiest to just create one big commit that changes all and be done with it? Or is the worry merge conflicts for open PRs?

In any case, if needed, how about a script that takes 10-20 from the stack of untreated packages (in alphabetical/random order), checks for which there are any open PRs and replaces those for which there are, and then runs prettier against those and pushes a PR for it?

damassi commented:
I think the concern is around the overall size of the repo, but perhaps it would work just fine! Worth a try.

damassi commented:
That wasn't so bad: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/35958.

alloy commented:
Hah, GitHub is having problems loading that page, and it isn't even the diff page. I’m sending their support a ping.

damassi commented:
Interesting, last night it would load just fine and could even browse a portion of the diff (I was very surprised). Maybe daytime server load?

alloy commented:
Seems to be because of CODEOWNERS http://twitter.com/mislav/status/1136281569868623874

sandersn commented:
@alloy I like your idea of a script. I think github and types-publisher are neither one able to scale to formatting every directory at once.

I am pretty short on time this week, but I created https://github.com/sandersn/dt-pr-prettier to hold the script for now. That'll give us a place to collaborate on something.

orta commented:
👋 alright folks love this idea, I'm gonna extend that script from @sandersn and get that so that we can run that pretty regularly over here - mainly because running prettier over everything would break ~160 PRs and most of those are probably drive-by PRs we'd never get responses for.

orta commented:
OK, given that an incremental approach to  adding this would take a_very_ long time to get this merged. 

To work around that I'm going to: 

- remove prettier from CI
- switch the husky config to use [precise-commits](https://github.com/nrwl/precise-commits#why-precise-commits)

This means, and code changes will have prettier running, and those that want to run prettier over their stuff can. Then we can manually start updating the most popular types in that script while no-one has active PRs

alloy commented:
That is probably going to surprise some contributors/maintainers, especially as style changes in the diff may overshadow the actual changes the contributor wanted to make. Not saying that this isn’t a fine decision, but wondering if there’s a need to give people a heads-up?

orta commented:
A message could be added to the PR template? but by that point you've already probably gone through the process

alloy commented:
True true. Well, I have no great (simple) suggestions... so let’s just do this?

orta commented:
Aye, I think so

alloy commented:
Sweet 🎉

damassi commented:
👋 @orta, ty  👍

vdh commented:
Are there any instructions on how to handle this new formatting requirement when attempting to make a PR? It creates a roadblock where for example someone previously configured the linting to require tabs (#27355), but Prettier is using spaces which then fails the lint check.

The "Formatting: Either use all tabs, or always use 4 spaces." instruction in the readme is not clear on what to do to resolve this tabs vs spaces situation now that Prettier and linting need to be definitively configured one way or the other so they don't conflict with each other.

sandersn commented:
@vdh The readme needs to be updated. In the meantime, tslints that require tabs need to have that rule removed.

sandersn commented:
#36595 updates the README

orta commented:
Maybe this is something we can run when I'm in town @sandersn ?

sandersn commented:
Sounds good to me.

orta commented:
We paired on this today and got pretty far, @eps1lon's script was enough to narrow it down to about 5k modules changed at once. https://github.com/orta/DefinitelyTyped/tree/test_new_prettier_code

Trying to apply prettier across all modules had a few flaws though, most notably syntax which doesn't have a config option means forcing particular libraries to only support specific typescript versions. e.g. syntax like

```ts
type MyThings = 
  | "one"
  | "two"
```

That format is a reasonably recent typescript addition, and we'd need to start bumping typescript versions for all `@types` modules if they are including prettier. Maybe they could be added to the prettier ignore? It's probably require some scripting, just going through the 'a's turned out a lot of these.

Another is that we needed to verify that they are still legit ts code - I found at least one reasonable  bug in the prettier parser ( https://github.com/prettier/prettier/issues/6634 ) but that was because I stopped in a very long stream of text output and managed to grab the filepath. There was more.

Getting `npm test` to work in this repo should verify the parsability, as there's a lot of intentionally broken TS code in this repo. There's probably more of these as the 

We think it'll be ok for it to generate new `@types/x` modules for every one of the changed (it'll back up the queue, but it'll get through eventually - but it would suck to send a message to every PR saying the publish is done so we added a check for `#no-publishing-comment` in the commit

I'm probably gonna hold off on carrying on with it, too many unknown unknowns for my time ATM but it's open for someone to continue, ideally give us a head up so that we can try get the PR in fast once its green. However, it's not just a simple `yarn prettier --write` 👯‍♂

orta commented:
Related, I'm going to remove precise commits and husky, https://github.com/DefinitelyTyped/DefinitelyTyped/pull/39018

We got pretty regular feedback that it was producing pretty weird code when adding small changes and regular contributors just learned to use `--no-verify`. I think style consistency is pretty important for the repo, but the trade-off on that style of incremental adoption just wasn't there.
